File: CodeBattleNetCore/Tetris/FigureRotator.cs
/*-
 * #%L
 * Codenjoy - it's a dojo-like platform from developers to developers.
 * %%
 * Copyright (C) 2018 Codenjoy
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */
using System;

namespace TetrisClient
{
	public static class FigureRotator
	{
		private static Func<Point, int, Point>[] ShiftDelegates = new Func<Point, int, Point>[]
		{
			(point, delta) => point.ShiftTop(delta),
			(point, delta) => point.ShiftRight(delta),
			(point, delta) => point.ShiftBottom(delta),
			(point, delta) => point.ShiftLeft(delta),
		};

		public static Point[] PredictCurrentFigurePoints(Rotation rotation, Point anchor, Element figureType)
		{
			var shiftTopAfterRotation = GetShiftAfterRotation(Direction.Up, rotation);
			var shiftRightAfterRotation = GetShiftAfterRotation(Direction.Right, rotation);
			var shiftBottomAfterRotation = GetShiftAfterRotation(Direction.Down, rotation);
			var shiftLeftAfterRotation = GetShiftAfterRotation(Direction.Left, rotation);

			switch (figureType)
			{
				case Element.BLUE:
					return new[]
					{
						anchor.Shift(shiftTopAfterRotation, 1),
						anchor,
						anchor.Shift(shiftBottomAfterRotation, 1),
						anchor.Shift(shiftBottomAfterRotation, 2),
					};
				case Element.CYAN:
					return new[]
					{
						anchor.Shift(shiftTopAfterRotation, 1),
						anchor,
						anchor.Shift(shiftBottomAfterRotation, 1),
						anchor.Shift(shiftBottomAfterRotation, 1)
							  .Shift(shiftLeftAfterRotation, 1),
					};
				case Element.ORANGE:
					return new[]
					{
						anchor.Shift(shiftTopAfterRotation, 1),
						anchor,
						anchor.Shift(shiftBottomAfterRotation, 1),
						anchor.Shift(shiftBottomAfterRotation, 1)
							  .Shift(shiftRightAfterRotation, 1),
					};
				case Element.YELLOW:
					return new[]
					{
						anchor,
						anchor.Shift(shiftRightAfterRotation, 1),
						anchor.Shift(shiftBottomAfterRotation, 1),
						anchor.Shift(shiftBottomAfterRotation, 1)
							  .Shift(shiftRightAfterRotation, 1),
					};
				case Element.GREEN:
					return new[]
					{
						anchor.Shift(shiftLeftAfterRotation, 1),
						anchor,
						anchor.Shift(shiftTopAfterRotation, 1),
						anchor.Shift(shiftTopAfterRotation, 1)
							  .Shift(shiftRightAfterRotation, 1),
					};
				case Element.PURPLE:
					return new[]
					{
						anchor.Shift(shiftLeftAfterRotation, 1),
						anchor,
						anchor.Shift(shiftRightAfterRotation, 1),
						anchor.Shift(shiftTopAfterRotation, 1),
					};
				case Element.RED:
					return new[]
					{
						anchor.Shift(shiftTopAfterRotation, 1)
							  .Shift(shiftLeftAfterRotation, 1),
						anchor.Shift(shiftTopAfterRotation, 1),
						anchor,
						anchor.Shift(shiftRightAfterRotation, 1),
					};
				default:
					throw new ArgumentException("Тип фигуры не поддерживается", nameof(figureType));
			}
		}

		private static Func<Point, int, Point> GetShiftAfterRotation(Direction shiftDirection, Rotation rotation)
		{
			return ShiftDelegates[((int)shiftDirection + (int)rotation) % 4];
		}

		private static Point Shift(this Point point, Func<Point, int, Point> shiftDelegate, int delta)
		{
			return shiftDelegate(point, delta);
		}
	}
}

File: CodeBattleJava/src/main/java/com/codenjoy/dojo/services/FigureRotator.java
package com.codenjoy.dojo.services;

import java.util.ArrayList;
import java.util.List;
import java.util.function.BiFunction;

import com.codenjoy.dojo.tetris.model.Elements;

import lombok.var;

public class FigureRotator {

    private static List<BiFunction<Point, Integer, Point>> SHIFT_DELEGATES = new ArrayList<>();
    static {
        SHIFT_DELEGATES.add((point, delta) -> point.shiftTop(delta));
        SHIFT_DELEGATES.add((point, delta) -> point.shiftRight(delta));
        SHIFT_DELEGATES.add((point, delta) -> point.shiftBottom(delta));
        SHIFT_DELEGATES.add((point, delta) -> point.shiftLeft(delta));
    };

    public static Point[] predictCurrentFigurePoints(Rotation rotation, Point anchor, Elements figureType) {
        var shiftTopAfterRotation = getShiftAfterRotation(Direction.UP, rotation);
        var shiftRightAfterRotation = getShiftAfterRotation(Direction.RIGHT, rotation);
        var shiftBottomAfterRotation = getShiftAfterRotation(Direction.DOWN, rotation);
        var shiftLeftAfterRotation = getShiftAfterRotation(Direction.LEFT, rotation);

        switch (figureType) {
        case BLUE:
            return new Point[] {
                    shift(anchor, shiftTopAfterRotation, 1),
                    anchor,
                    shift(anchor, shiftBottomAfterRotation, 1),
                    shift(anchor, shiftBottomAfterRotation, 2),
            };
        case CYAN:
            return new Point[] {
                    shift(anchor, shiftTopAfterRotation, 1),
                    anchor,
                    shift(anchor, shiftBottomAfterRotation, 1),
                    shift(shift(anchor, shiftBottomAfterRotation, 1), shiftLeftAfterRotation, 1),
            };
        case ORANGE:
            return new Point[] {
                    shift(anchor, shiftTopAfterRotation, 1),
                    anchor,
                    shift(anchor, shiftBottomAfterRotation, 1),
                    shift(shift(anchor, shiftBottomAfterRotation, 1), shiftRightAfterRotation, 1),
            };
        case YELLOW:
            return new Point[] {
                    anchor,
                    shift(anchor, shiftRightAfterRotation, 1),
                    shift(anchor, shiftBottomAfterRotation, 1),
                    shift(shift(anchor, shiftBottomAfterRotation, 1), shiftRightAfterRotation, 1),
            };
        case GREEN:
            return new Point[] {
                    shift(anchor, shiftLeftAfterRotation, 1),
                    anchor,
                    shift(anchor, shiftTopAfterRotation, 1),
                    shift(shift(anchor, shiftTopAfterRotation, 1), shiftRightAfterRotation, 1),
            };
        case PURPLE:
            return new Point[] {
                    shift(anchor, shiftLeftAfterRotation, 1),
                    anchor,
                    shift(anchor, shiftRightAfterRotation, 1),
                    shift(anchor, shiftTopAfterRotation, 1),
            };
        case RED:
            return new Point[] {
                    shift(shift(anchor, shiftTopAfterRotation, 1), shiftLeftAfterRotation, 1),
                    shift(anchor, shiftTopAfterRotation, 1),
                    anchor,
                    shift(anchor, shiftRightAfterRotation, 1),
            };
        default:
            throw new IllegalArgumentException("Тип фигуры не поддерживается: " + figureType);
        }
    }

    private static BiFunction<Point, Integer, Point> getShiftAfterRotation(Direction shiftDirection,
            Rotation rotation) {
        return SHIFT_DELEGATES.get((shiftDirection.ordinal() + rotation.ordinal()) % 4);
    }

    private static Point shift(Point point, BiFunction<Point, Integer, Point> shiftDelegate, int delta) {
        return shiftDelegate.apply(point, delta);
    }

}
